### PR TITLE
feat: parameter defaults, make structs const-constructible

### PR DESCRIPTION
**Pull request by Desdaemon**
## Changes

Part of #960.

- `#[frb(default = ..)]` to denote parameter defaults
  - [x] Functions
  - [x] Struct constructors
  - [x] Enum constructors
- Classes with all final fields are now const-constructible

## Checklist

- [x] An issue to be fixed by this PR is listed above.
- [x] New tests are added to ensure new features are working. End-to-end tests are usually in the `./frb_example/pure_dart` example, more specifically, `rust/src/api.rs` and `dart/lib/main.dart`.
- [x] The code generator is run and the code is formatted (via `just precommit`).
- [x] If this PR adds/changes features, documentations (in the `./book` folder) are updated.
- [x] CI is passing.

## Remark for PR creator

- Justfile is a task runner for the command line interface that allows you to run shell commands from a file named `justfile` in your project directory. You can use Justfile after [installing it](https://github.com/casey/just). Note that commands written in `justfile` of this repository are expected to be run in `bash`, not `cmd` or `powershell`. Running `just ...` commands in `cmd` or `powershell` will produce errors as the syntax is not compatible. On Windows, you can use `git bash` if you have Git installed.
- If fzyzcjy does not reply for a few days, maybe he just did not see it, so please ping him.

### CODE REVIEW

**Comment by Desdaemon** [comment]
I was debating whether to support mutable enums using Freezed's `@unfreezed`, but it might be more appropriate to implement that in a separate PR.

**Comment by fzyzcjy** [comment]
Good job :) I will review probably today (btw one CI is failing)

File: frb_codegen/src/ir/field.rs [hunk 1]
```diff
@@ -1,9 +1,46 @@
...
+            ""
+        } else {
+            self.ty.dart_required_modifier()
+        }
+    }
+    pub fn field_default(&self, freezed: bool) -> String {
+        self.default
+            .as_ref()
+            .map(|default| {
+                let default = match default {
```
Comment (fzyzcjy) [inline frb_codegen/src/ir/field.rs:25]: Not sure whether `default` will be a keyword in the future? https://users.rust-lang.org/t/what-is-the-default-keyword-means-in-rust/57448

**Review by fzyzcjy — COMMENTED** [review]
Good job! Only two nits

**Review by fzyzcjy — APPROVED** [review]
(mark as approve, thus merge after the nits are solved)

File: frb_codegen/src/ir/field.rs (hunk 1 above)
Comment (Desdaemon) [reply frb_codegen/src/ir/field.rs:25]: IIRC it was meant to be used for specialization, as in `default fn`. I'm welcome to suggestions for replacements however.

Comment (fzyzcjy) [reply frb_codegen/src/ir/field.rs:25]: Then maybe we just keep it like this. If in the future it becomes a keyword, we can `r#` at that time easily.

**Comment by Desdaemon** [comment]
@fzyzcjy Good to merge now

**Comment by fzyzcjy** [comment]
Great!

**Comment by fzyzcjy** [comment]
https://github.com/fzyzcjy/flutter_rust_bridge/actions/runs/4309797891/jobs/7517592544

hmm CI in master fails

**Comment by Desdaemon** [comment]
> https://github.com/fzyzcjy/flutter_rust_bridge/actions/runs/4309797891/jobs/7517592544
> 
> hmm CI in master fails

That doesn't sound right... Let me reopen another PR.

**Pull request merged**